backend/services/user_groups.go

```diff
@@ -0,0 +1,421 @@
+	remainingItemsCount := (groupLength - int(prevLastIndex)) % int(pagination.PageSize)
+
+	currLastIndex := int(pagination.Page * pagination.PageSize)
+	pageSize := pagination.PageSize
+	if pagination.Page == int64(totalPages) {
+		currLastIndex = int(prevLastIndex) + remainingItemsCount
+		pageSize = int64(remainingItemsCount)
+	}
+
+	paginatedResults := userGroupsDTO[prevLastIndex:currLastIndex]
+	paginatedData := &dtos.PaginationWrapper{
+		PageNumber: pagination.Page,
+		PageSize:   pageSize,
+		Content:    paginatedResults,
+		TotalCount: int64(groupLength),
+		TotalPages: int64(totalPages),
+	}
+	return paginatedData, nil
+}
+
+// Lists all user groups for an operation; op admins and sys admins can view
+func ListUserGroupsForOperation(ctx context.Context, db *database.Connection, i ListUserGroupsForOperationInput) ([]*dtos.UserGroupOperationRole, error) {
+	operation, err := lookupOperation(db, i.OperationSlug)
+	if err := policyRequireWithAdminBypass(ctx, policy.CanListUserGroupsOfOperation{OperationID: operation.ID}); err != nil {
+		return nil, backend.WrapError("Unwilling to list usergroups", backend.UnauthorizedReadErr(err))
+	}
+
+	query := sq.Select("slug", "name", "role").
+		From("user_group_operation_permissions").
+		LeftJoin("user_groups ON user_group_operation_permissions.group_id = user_groups.id").
+		Where(sq.Eq{"operation_id": operation.ID, "user_groups.deleted_at": nil}).
+		OrderBy("user_group_operation_permissions.created_at ASC")
+
+	i.UserGroupFilter.AddWhere(&query)
+
+	var userGroups []userGroupAndRole
+	err = db.Select(&userGroups, query)
+	if err != nil {
+		return nil, backend.WrapError("Cannot list user groups for operation", backend.DatabaseErr(err))
+	}
+	userGroupsDTO := wrapListUserGroupsForOperationResponse(userGroups)
+	return userGroupsDTO, nil
+}
+
+func wrapListUserGroupsForOperationResponse(userGroups []userGroupAndRole) []*dtos.UserGroupOperationRole {
+	userGroupsDTO := make([]*dtos.UserGroupOperationRole, len(userGroups))
+	for idx, userGroup := range userGroups {
+		userGroupsDTO[idx] = &dtos.UserGroupOperationRole{
+			UserGroup: dtos.UserGroupAdminView{
+				Slug: userGroup.Slug,
+				Name: userGroup.Name,
+			},
+			Role: userGroup.Role,
+		}
+	}
+	return userGroupsDTO
+}
+
+// lists all user groups that can be added to an operation
+// no pagination, because this is used for the search bar
+func ListUserGroups(ctx context.Context, db *database.Connection, i ListUserGroupsInput) ([]*dtos.UserGroupAdminView, error) {
+	operation, err := lookupOperation(db, i.OperationSlug)
+	if err := policyRequireWithAdminBypass(ctx, policy.CanListUserGroupsOfOperation{OperationID: operation.ID}); err != nil {
+		return nil, backend.WrapError("Unwilling to list usergroups", backend.UnauthorizedReadErr(err))
+	}
+
+	if strings.ContainsAny(i.Query, "%_") || strings.TrimFunc(i.Query, unicode.IsSpace) == "" {
+		return []*dtos.UserGroupAdminView{}, nil
+	}
```

Comment:
Do we actually need this? To me, it looks like we're effectively saying that a group with an underscore isn't findable. I don't know if there are other rules that preclude that possibility, but if we _can_ support it, it seems like we should.

One suggestion: maybe instead of removing % and _, you could instead use `strings.ReplaceAll(i.Query, "%", "\\%)` (ditto with underscore). I think that'll properly escape them.

In any event, this is more fine tuning than an actual issue as far as I'm concerned.

Reply:
yeah, I included this because there's a similar check for users (users.go 324):
```go
func ListUsers(ctx context.Context, db *database.Connection, i ListUsersInput) ([]*dtos.User, error) {
	if strings.ContainsAny(i.Query, "%_") || strings.TrimFunc(i.Query, unicode.IsSpace) == "" {
		return []*dtos.User{}, nil
	}
```
but I wasn't really sure if it was a security issue or what, so I'm down for removing it if it seems good to you as well

Reply:
I don't _think_ it's a security issue. those two characters are just placeholders in like queries -- % for  effectively `.*` and _ for effectively `.` (in regex speak). They could mess up the query -- which is why I was proposing that we escape them instead. I think that % and _ are unlikely username characters, but they might be more likely group name characters...

I'm now inclined to leave it as is just because we're already doing this somewhere else. Maybe a separate effort is done to see if we can remove those characters (or if we can even create groups with those characters)

Reply:
sounds good!